pkg/providers/openai/openaiclient.go

```diff
@@ -0,0 +1,172 @@
+// TODO: Explore resource pooling
+// TODO: Optimize Type use
+// TODO: Explore Hertz TLS & resource pooling
+// OpenAI package provide a set of functions to interact with the OpenAI API.
+package openai
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+	"os"
+
+	"gopkg.in/yaml.v2"
+
+	"glide/pkg/providers"
+
+	"github.com/cloudwego/hertz/pkg/app/client"
+)
+
+const (
+	defaultBaseURL      = "https://api.openai.com/v1"
+	defaultOrganization = ""
+)
+
+// ErrEmptyResponse is returned when the OpenAI API returns an empty response.
+var (
+	ErrEmptyResponse = errors.New("empty response")
+	requestBody      struct {
+		Message []struct {
+			Role    string `json:"role"`
+			Content string `json:"content"`
+		} `json:"message"`
+		MessageHistory []string `json:"messageHistory"`
+	}
+)
+
+type APIType string
+
+const (
+	APITypeOpenAI  APIType = "OPEN_AI"
+	APITypeAzure   APIType = "AZURE"
+	APITypeAzureAD APIType = "AZURE_AD"
+)
+
+// Client is a client for the OpenAI API.
+type Client struct {
+	Provider     providers.Provider
+	PoolName     string
+	baseURL      string
+	payload      []byte
+	organization string
+	apiType      APIType
+	httpClient   *client.Client
+
+	// required when APIType is APITypeAzure or APITypeAzureAD
+	apiVersion string
+}
+
+// OpenAiClient creates a new client for the OpenAI API.
+//
+// Parameters:
+// - poolName: The name of the pool to connect to.
+// - modelName: The name of the model to use.
+//
+// Returns:
+// - *Client: A pointer to the created client.
+// - error: An error if the client creation failed.
+func OpenAiClient(poolName string, modelName string, payload []byte) (*Client, error) {
+	providerName := "openai"
+
+	// Read the YAML file
+	data, err := os.ReadFile("/Users/max/code/Glide/config.yaml")
```

Comment:
@roma-glushko How should I plan on accessing the config provided by the user?

Reply:
@mkrueger12 I'm going to add a struct that will hold all configs and that instance is going to be used in Router struct to init pools and models in them. Provider client instances are going to be created according to the config. I'm considering that as a part of the task related to introducing pools.

Reply:
Ok so the provider clients basically just need to take that struct and build the POST request to the API endpoint?

Reply:
@mkrueger12 the provider client is going to be inited with these configs and then it will take only the unified request schema and combine these two to make an actual request to LLM provider API, receive a response, unify it and then return it all the way back to the user/Glide client.

Reply:
So does the incoming request gets deserialized into a struct representing the unified request schema? Is the following what the request schema looks like? If not can you help illustrate that for me?

```json
{
  "message": [
      {
        "role": "system",
        "content": "You are a helpful assistant."
      },
      {
        "role": "user",
        "content": "Where was it played?"
      }
    ],
  "messageHistory": [
    {"role": "system", "content": "You are a helpful assistant."},
    {"role": "user", "content": "Who won the world series in 2020?"},
    {"role": "assistant", "content": "The Los Angeles Dodgers won the World Series in 2020."},
  ]
}
```

Reply:
Actually to really looks something like this with params being optional (or we don't allow them now):

```json
{ 
  "provider": "openai",
  "model": "gpt-3.5-turbo",
  "params": {"temperature": 0.7, "topP": 1.0, "frequencyPenalty": 0.0, "presencePenalty": 0.0, "max_tokens": 256, "n": 1},
  "message": "tell me a joke",
  "messageHistory": ["Hello there", "How are you?", "I'm good, how about you?"]
}
```

Reply:
```go

type UnifiedAPIData struct {
	Provider       string            `json:"provider"`
	Model          string            `json:"model"`
	Params         map[string]float64 `json:"params"`
	Message        string            `json:"message"`
	MessageHistory []string          `json:"messageHistory"`
}

```

Reply:
So the incoming request gets deserialized into a struct representing the unified request schema? Is the following what the unified schema looks like? If not can you help illustrate that for me?

```json
{
  "provider": "openai",
  "model": "gpt-3.5-turbo",
  "params": {"temperature": 0.7, "topP": 1.0, "frequencyPenalty": 0.0, "presencePenalty": 0.0, "max_tokens": 256, "n": 1},
  "message": [
      {
        "role": "system",
        "content": "You are a helpful assistant."
      },
      {
        "role": "user",
        "content": "Where was it played?"
      }
    ],
  "messageHistory": [
    {"role": "system", "content": "You are a helpful assistant."},
    {"role": "user", "content": "Who won the world series in 2020?"},
    {"role": "assistant", "content": "The Los Angeles Dodgers won the World Series in 2020."},
  ]
}
```

Reply:
```go

type UnifiedAPIData struct {
	Provider       string            `json:"provider"`
	Model          string            `json:"model"`
	Params         map[string]float64 `json:"params"`
	Message        string            `json:"message"`
	MessageHistory []string          `json:"messageHistory"`
}

```

Reply:
Should message and messageHistory be bytes?

Reply:
> the incoming request gets deserialized into a struct representing the unified request schema?
@mkrueger12 yes, that was the idea. The that struct instance is passed further across the system until the request is fulfilled. 

>  Is the following what the unified schema looks like? If not can you help illustrate that for me?

It does partially. We call the schema unified because it's not connected to any provider but could be handled by any of them if that makes sense. So model & param are configured on the model pool level (and only there for the time being until we allow to override them somehow). That leaves us with `message` & `messageHistory`.

Both fields look pretty the same in terms of structure e.g. message is essentially the last item in the general message history. So we could even go further and concat them if that makes sense from the usage & mapping to concreate provider schemas.